### ISSUE TITLE
[RFC] Performance monitoring

### ISSUE THREAD

**Issue by theofidry** [infection/infection#1176]
Given #1173, the more frequent recent works on performances, I think it is time to step up our game. Just for a bit more context, here is the current situation:

- we profile occasionally in our corners each one on different projects
- profiling is not necessarily easy as we have a lot of "noisy" or unreliable parts of the applications (in terms of profiling), e.g.:
  - a lot of I/O operations: read files, dump files
  - process handling: for the tests, for the mutation analysis
  - parsing of PHP files into AST and traversing: it's often a dominant part of the profile yet not something we can do much about

What I would like to achieve is an easier experience to profile, provide more reliable and robust results.

In order to achieve that, I think there is several actionable:

- Create dedicated small cases, e.g.:
  - the initial test run process creation (no execution): this would allow to profile the command creation
  - file parsing and traverse to get mutations (this is mostly PhpParser core and a bit of our stuff)
  - mutation -> mutant process creation which includes the coverage parsing + creation of the framework config file
  - the report creation
- Set up some benchmarks: with the few cases above it would be easier to have a real benchmark, not profile, e.g. via https://github.com/phpbench/phpbench
- A dry-run mode: i.e. a mode where we skip the process handling and reporting (which is a huge part of instability due to the processes and usleep calls). This would be an ideal complement with the `--coverage` option
- A utility to ease profiling: there is often a few recurring points to do for a profile, e.g.:
  - make sure xdebug is disabled (huge overhead)
  - make sure pcov is disabled (skews blackfire giving weird profiles)
  - make sure the autoloader is dumped in authoritative mode
  - make sure the dependencies are up to date (easy to forget if you check out two different versions)
  - generate the coverage, make sure that the infection command includes `--quiet` for example
- Request an OSS license for Blackfire? (not sure it's doable, but hey we can ask)

I don't expect all the above to solve 100% of our problems or to find big performance improvements from any of them. Also some might be too hard to implement in the current form. For example, wouldn't it be amazing that at each PR for all the scenarios above are compared with both versions to get the difference?

In other words, my goal here is to list some actionnables in order to climb the ladder. I do not think we need to try to reach the top of it right away. For example having just a few small scenarios which are more "profile friendly" or a dry-run mode would already be a huge step forward IMO.

**Comment by theofidry** [comment]
Note from @sanmai, includes #1150 as well

**Comment by theofidry** [comment]
I added two scenarios easy to profile in #1178. Feel free to suggest more scenarios that you think are worth monitoring/profiling
